components/net/filemanager_thread.rs

```diff
@@ -87,56 +89,93 @@ impl<UI: 'static + UIProvider> FileManagerThreadFactory<UI> for IpcSender<FileMa
     }
 }
 
+struct FileStoreEntry {
+    /// Origin of the entry's "creator"
+    origin: String,
+    /// Kind of entry
+    kind: FileStoreEntryKind,
+    /// Reference counting
+    refs: usize,
+}
+
+enum FileStoreEntryKind {
+    /// PathOnly consists of filepath, optional slice position, and origin of requester
+    PathOnly(PathBuf),
+    /// Buffered entry which contains bytes, optional slice position, and origin of requester
+    Buffered(BlobURLStoreEntry),
+    /// Indirect entry
+    Indirect(Uuid, RelativePos),
+}
+
 struct FileManager<UI: 'static + UIProvider> {
     receiver: IpcReceiver<FileManagerThreadMsg>,
-    idmap: HashMap<Uuid, PathBuf>,
+    store: Arc<RwLock<HashMap<Uuid, FileStoreEntry>>>,
```

Comment:
I think this can be handled later. I'd prefer if there was a nice concurrent hashmap out there (there probably is) that can be used to abstract over the downgrading.